File: .claude/hooks/pre-webfetch-html-to-markdown.mjs
#!/usr/bin/env node
// @hookstack pre-webfetch-html-to-markdown
// Convertit les pages HTML en Markdown avant traitement WebFetch (PreToolUse WebFetch)
import { readFileSync } from 'node:fs';
import { execSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const MAX_CHARS = 30_000;

function defaultFetchUrl(url) {
  return execSync('curl -sL --max-time 10 --user-agent "Mozilla/5.0" "$HOOK_URL"', {
    encoding: 'utf8',
    timeout: 15_000,
    env: { ...process.env, HOOK_URL: url },
  }).trim();
}

function defaultConvertHtml(html) {
  return execSync('pandoc -f html -t markdown --wrap=none', {
    input: html,
    encoding: 'utf8',
    timeout: 10_000,
  }).trim();
}

function defaultHasPandoc() {
  try { execSync('which pandoc', { encoding: 'utf8', timeout: 2_000 }); return true; } catch { return false; }
}

function stripHtml(html) {
  return html
    .replace(/<script[\s\S]*?<\/script>/gi, '')
    .replace(/<style[\s\S]*?<\/style>/gi, '')
    .replace(/<[^>]+>/g, ' ')
    .replace(/&amp;/g, '&').replace(/&lt;/g, '<').replace(/&gt;/g, '>').replace(/&quot;/g, '"').replace(/&#39;/g, "'").replace(/&nbsp;/g, ' ')
    .replace(/[ \t]{2,}/g, ' ')
    .replace(/\n{3,}/g, '\n\n')
    .trim();
}

export function run(input, {
  fetchUrl = defaultFetchUrl,
  convertHtml = defaultConvertHtml,
  hasPandoc = defaultHasPandoc,
} = {}) {
  if (input.tool_name !== 'WebFetch') return null;

  const url = input.tool_input?.url ?? '';
  if (!url || !/^https?:\/\//i.test(url)) return null;

  let html;
  try { html = fetchUrl(url); } catch { return null; }
  if (!html?.trim()) return null;

  // Uniquement les pages HTML — laisser passer JSON, binaires, etc.
  if (!/<html|<!doctype\s+html/i.test(html.slice(0, 2000))) return null;

  let markdown;
  try {
    markdown = hasPandoc() ? convertHtml(html) : stripHtml(html);
  } catch {
    try { markdown = stripHtml(html); } catch { return null; }
  }

  if (!markdown?.trim()) return null;

  let content = markdown.trim();
  let truncated = false;
  if (content.length > MAX_CHARS) {
    content = content.slice(0, MAX_CHARS);
    truncated = true;
  }

  let domain;
  try { domain = new URL(url).hostname; } catch { domain = url; }
  const suffix = truncated ? ` (truncated to ${MAX_CHARS} chars)` : '';

  return {
    decision: 'block',
    reason: `[webfetch-html-to-markdown] \`${domain}\` converted to Markdown${suffix}:\n\n${content}`,
  };
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const input = JSON.parse(readFileSync(0, 'utf8'));
  const result = run(input);
  if (result) process.stdout.write(JSON.stringify(result));
}

File: .claude/hooks/pre-webfetch-html-to-markdown.py
#!/usr/bin/env python3
# @hookstack pre-webfetch-html-to-markdown
"""Converts HTML pages to Markdown before WebFetch processing (PreToolUse WebFetch)."""
import json
import re
import subprocess
import sys
from urllib.parse import urlparse

MAX_CHARS = 30_000


def _fetch_url(url):
    return subprocess.run(
        f'curl -sL --max-time 10 --user-agent "Mozilla/5.0" "$HOOK_URL"',
        shell=True,
        capture_output=True,
        text=True,
        timeout=15,
        check=True,
        env={**__import__("os").environ, "HOOK_URL": url},
    ).stdout.strip()


def _convert_html(html):
    return subprocess.run(
        "pandoc -f html -t markdown --wrap=none",
        shell=True,
        capture_output=True,
        text=True,
        timeout=10,
        check=True,
        input=html,
    ).stdout.strip()


def _has_pandoc():
    try:
        subprocess.run(
            "which pandoc", shell=True, capture_output=True, text=True, timeout=2, check=True
        )
        return True
    except Exception:
        return False


def strip_html(html):
    out = re.sub(r"<script[\s\S]*?</script>", "", html, flags=re.IGNORECASE)
    out = re.sub(r"<style[\s\S]*?</style>", "", out, flags=re.IGNORECASE)
    out = re.sub(r"<[^>]+>", " ", out)
    out = out.replace("&amp;", "&")
    out = out.replace("&lt;", "<")
    out = out.replace("&gt;", ">")
    out = out.replace("&quot;", '"')
    out = out.replace("&#39;", "'")
    out = out.replace("&nbsp;", " ")
    out = re.sub(r"[ \t]{2,}", " ", out)
    out = re.sub(r"\n{3,}", "\n\n", out)
    return out.strip()


def run(input_data, *, fetch_url=None, convert_html=None, has_pandoc=None):
    if fetch_url is None:
        fetch_url = _fetch_url
    if convert_html is None:
        convert_html = _convert_html
    if has_pandoc is None:
        has_pandoc = _has_pandoc

    if input_data.get("tool_name") != "WebFetch":
        return None

    url = (input_data.get("tool_input") or {}).get("url") or ""
    if not url or not re.match(r"^https?://", url, re.IGNORECASE):
        return None

    try:
        html = fetch_url(url)
    except Exception:
        return None
    if not (html or "").strip():
        return None

    # Uniquement les pages HTML — laisser passer JSON, binaires, etc.
    if not re.search(r"<html|<!doctype\s+html", html[:2000], re.IGNORECASE):
        return None

    try:
        markdown = convert_html(html) if has_pandoc() else strip_html(html)
    except Exception:
        try:
            markdown = strip_html(html)
        except Exception:
            return None

    if not (markdown or "").strip():
        return None

    content = markdown.strip()
    truncated = False
    if len(content) > MAX_CHARS:
        content = content[:MAX_CHARS]
        truncated = True

    try:
        domain = urlparse(url).hostname or url
    except Exception:
        domain = url
    suffix = f" (truncated to {MAX_CHARS} chars)" if truncated else ""

    return {
        "decision": "block",
        "reason": f"[webfetch-html-to-markdown] `{domain}` converted to Markdown{suffix}:\n\n{content}",
    }


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result:
        sys.stdout.write(json.dumps(result))
